src/Math/Calculator.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace Drupal\rules\Math;
+
+use Drupal\rules\Math\Exception\IncorrectExpressionException;
+use Drupal\rules\Math\Token\ConstantToken;
+use Drupal\rules\Math\Token\FunctionToken;
+use Drupal\rules\Math\Token\NumberToken;
+use Drupal\rules\Math\Token\OperatorTokenInterface;
+
+class Calculator {
+
+  /**
+   * Static cache of token streams in reverse polish (postfix) notation.
+   *
+   * @var array
+   */
+  protected $tokenCache = [];
+
+  public function __construct() {
+    $this->lexer = new Lexer();
+
+    $this->lexer
+      ->addOperator('\+', 'Drupal\rules\Math\Token\PlusToken')
+      ->addOperator('\-', 'Drupal\rules\Math\Token\MinusToken')
+      ->addOperator('\*', 'Drupal\rules\Math\Token\MultiplyToken')
+      ->addOperator('\/', 'Drupal\rules\Math\Token\DivisionToken')
+      ->addOperator('\%', 'Drupal\rules\Math\Token\ModulusToken')
+      ->addOperator('\^', 'Drupal\rules\Math\Token\PowerToken');
+
+    $this->lexer
+      ->addFunction('abs', 'abs', 1)
+      ->addFunction('acos', 'acos', 1)
+      ->addFunction('acosh', 'acosh', 1)
+      ->addFunction('asin', 'asin', 1)
+      ->addFunction('asinh', 'asinh', 1)
+      ->addFunction('atan2', 'atan2', 2)
+      ->addFunction('atan', 'atan', 1)
+      ->addFunction('atanh', 'atanh', 1)
+      ->addFunction('ceil', 'ceil', 1)
+      ->addFunction('cos', 'cos', 1)
+      ->addFunction('cosh', 'cosh', 1)
+      ->addFunction('deg2rad', 'deg2rad', 1)
+      ->addFunction('exp', 'exp', 1)
+      ->addFunction('floor', 'floor', 1)
+      ->addFunction('hypot', 'hypot', 2)
+      ->addFunction('log10', 'log10', 1)
+      ->addFunction('log', 'log', 2)
+      ->addFunction('max', 'max', 2)
+      ->addFunction('min', 'min', 2)
+      ->addFunction('pow', 'pow', 2)
+      ->addFunction('rad2deg', 'rad2deg', 1)
+      ->addFunction('rand', 'rand', 2)
+      ->addFunction('round', 'round', 1)
+      ->addFunction('sin', 'sin', 1)
+      ->addFunction('sinh', 'sinh', 1)
+      ->addFunction('sqrt', 'sqrt', 1)
+      ->addFunction('tan', 'tan', 1)
+      ->addFunction('tanh', 'tanh', 1);
+
+    $this->lexer->addConstant('pi', pi());
+    $this->lexer->addConstant('e', exp(1));
+  }
+
+  public function calculate($expression, $variables) {
+    $hash = md5($expression);
+    if (isset($this->tokenCache[$hash])) {
+      return $this->tokenCache[$hash];
+    }
+
+    $stream = $this->lexer->tokenize($expression);
+    $tokens = $this->lexer->postfix($stream);
```

Comment:
So after the postfix, tokens will be in reverse polish notation. Which is the reason why you can use a stack to evaluate. Would be cool to explain this a bit more